contributors/design-proposals/storage/pv-to-rbd-mapping.md

```diff
@@ -0,0 +1,119 @@
+# RBD Volume to PV Mapping 
+
+Authors: krmayankk@
+
+### Problem
+
+The RBD Dynamic Provisioner currently generates rbd volume names which are random. 
+The current implementation generates a UUID and the rbd image name becomes 
+image := fmt.Sprintf("kubernetes-dynamic-pvc-%s", uuid.NewUUID()). This RBD image 
+name is stored in the PV. The PV also has a reference to the PVC to which it binds. 
+The problem with this approach is that if there is a catastrophic etcd data loss 
+and all PV's are gone, there is no way to recover the mapping from RBD to PVC. The
+RBD volumes for the customer still exist, but we have no way to tell which rbd 
+volumes belong to which customer.
+
+## Goal
+We want to store some information about the PVC in RBD image name/metadata, so that 
+in catastrophic situations, we can derive the PVC name from rbd image name/metadata
+and allow customer the following options:
+- Backup RBD volume data for specific customers and hand them their copy before deleting 
+  the RBD volume. Without knowing from rbd image name/metadata, which customers they 
+  belong to we cannot hand those customers their data.
+- Create PV with the given RBD name and pre-bind it to the desired PVC so that customer
+  can get its data back.
+
+## Non Goals
+This proposal doesnt attempt to undermine the importance of etcd backups to restore
+data in catastrophic situations. This is one additional line of defense in case our
+backups are not working.
+
+## Motivation
+
+We recently had an etcd data loss which resulted in loss of this rbd to pv mapping 
+and there was no way to restore customer data. This proposal aims to store pvc name 
+as metadata in the RBD image so that in catostrophic scenarios, the mapping can be 
+restored by just looking at the RBD's.
+
+## Current Implementation
+
+```go
+func (r *rbdVolumeProvisioner) Provision() (*v1.PersistentVolume, error) {
+...
+
+ // create random image name
+ image := fmt.Sprintf("kubernetes-dynamic-pvc-%s", uuid.NewUUID())
+ r.rbdMounter.Image = image
+```
+
+## Proposal 1
+
+Make the RBD Image name as base64 encoded PVC name(namespace+name)
+
+```go
+import b64 "encoding/base64"
+...
+
+
+func (r *rbdVolumeProvisioner) Provision() (*v1.PersistentVolume, error) {
+...
+
+
+ // create RBD Image name 
+ image := fmt.Sprintf("kubernetes-dynamic-pvc-%s", uuid.NewUUID())
+
+ // Create a base64 encoding of the PVC Namespace and Name
+ rbdImageName := b64.StdEncoding.EncodeToString([]byte(r.options.PVC.Name+"-"+r.options.PVC.Namespace))
```

Comment:
`-` is not a delimiter, it doesn't help avoid collision. Try
`r.options.PVC.Namespace+"/"+r.options.PVC.Name`

Reply:
again, this will break old krbd where image length cannot be over 95 bytes.

Reply:
fixed and moved this proposal at the bottom